### ISSUE TITLE
Support white-listing of locations

### ISSUE THREAD

**Issue by stechu** [open-web3-stack/open-runtime-module-library#703]
One common error in XCM is when sending to unsupported location. It would be great for `xtoken` to provide an optional support for white-listing locations.

**Comment by xlc** [comment]
We could have a location filter, and another pallet to manage the whitelist registration

**Comment by stechu** [comment]
> We could have a location filter, and another pallet to manage the whitelist registration

yeah, adding interface for hooking the other pallet that manage the whitelist registration and then `xtoken` can return an `UnsupportedLocation` error rather than returning `XCMExecutionFailed` would be nice.

**Comment by Dengjianping** [comment]
Yeah, some teams are using hardcoded locations, some are using a on-chain manager to manage locations.

If `xtoken` is going to support getting  location by manager, we can add an associated type to `Config`, just like:
```rust
#[pallet::config]
pub trait Config: frame_system::Config {
    // ...
    type WhiteListingMultiLocations: WhiteListingMultiLocations<Self::CurrencyId>;
    // ...
}
```

And this trait `WhiteListingMultiLocations` should include methods like get supported location by `CurrencyId`.
```rust
pub trait WhiteListingMultiLocations<CurrencyId, Balance> {
    fn get_supported_locations(&self, currency_id: CurrencyId) -> Option<MultiLocation>;
    fn get_all_supported_locations(&self) -> Option<Vec<MultiLocation>>;
    fn get_xcm_fee(&self, currency_id: CurrencyId) -> Balance;
    // fn get_asset(&self, currency_id: CurrencyId) -> AssetInfo;
    // or more methods 
}
```

**Comment by xlc** [comment]
PRs are welcome.

**Comment by xlc** [comment]
closed by #714
